### DISCUSSION TITLE
pre-configuration of well-known decorators

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/pyright#4129] [Ideas]
This is a pattern that is common in my own code, and is obvious to a minimally experienced reader, but is opaque to pyright:

```!python
from dataclasses import dataclass

immutable = dataclass(frozen=True)
del dataclass  # guard against unintended use, below

@immutable
class C:
  x: int
```

In this case, pyright currently (1.1.278) reports `reportUninitializedInstanceVariable: Instance variable "x" is not initialized in the class body or __init__ method (pyright)`.

First, let me agree:

1. this is an uncommon and unimportant use case
2. any work involved to "improve" the above situation has cost outweighing the benefit.

 
That said, would you entertain discussion of hypotheticals? I find it useful and informative to discuss ideals apart from implementation.
 
How _could_ a type checker (pyright or otherwise) handle this situation in a way that matches runtime behavior and readers' intuition?

Based on the behavior alone, it smells to me like there is special-case code inside pyright that is triggered for objects named `dataclasses.dataclass`. Is that the case? That would explain why aliases would confound the analysis.

**Comment by erictraut** [comment]
Yes, there is necessarily special-case code inside pyright (and other Python type checkers) to handle `dataclass`. This also applies to other classes decorated with a "dataclass transform" (see PEP 681). This is necessary because `dataclass` includes parameters like `frozen`, `init` and `kw_only` that affect typing behaviors. The keyword arguments provided for these parameters must be static bool expressions, and it must be possible to analyze them in context where the `dataclass` decorator is applied (i.e. at the point where it is used as a decorator). Otherwise these arguments are lost.

I can offer a couple of suggestions for making this work:

1. Don't define a separate symbol for `immutable`. Simply use `dataclass(frozen=True)` in place. This is how `dataclass` is demonstrated in the samples provided in PEP 557 (which introduced `dataclass`), so it's the way the authors intended for it to be used. You're deviating from this intended usage patter, and that's getting you into some trouble.

2. You can use the new PEP 681 "dataclass transform" mechanism to fool the type checker into believing that `immutable` should act like `dataclass`. There's no way to provide the `frozen` parameter in this case, so the type checker won't detect any violations related to immutability, so I would still recommend the first option over this one.

```python
from typing import TYPE_CHECKING, TypeVar, dataclass_transform

if TYPE_CHECKING:
    T = TypeVar("T")
    @dataclass_transform()
    def immutable(kls: type[T]) -> type[T]: ...
else:
    from dataclasses import dataclass
    immutable = dataclass(frozen=True)
    del dataclass
```
